### PR TITLE
Documentation about how to write code documentation

### PR DESCRIPTION
**Pull request by qianqianzhu**
### Description

This PR adds a documentation about how to write code documentation.
It is related to issues: [#812](https://github.com/marian-nmt/marian-dev/issues/812). 
It also fixes several Doxygen and Sphinx warnings related to [#811](https://github.com/marian-nmt/marian-dev/issues/811)

List of changes:
- add a new documentation about how to write code documentation
- fix warnings from Exhale (duplicated signatures)
- fix warnings in markdown files (e.g. operators.md shows "None:any reference target not found")
- fix math formula cannot be well displayed (e.g., [marian::affine](https://marian-nmt.github.io/docs/api/api/function_group__graph__ops_1ga8a5bcad46fab763303f0e576bbb4f33e.html#exhale-function-group-graph-ops-1ga8a5bcad46fab763303f0e576bbb4f33e))

Added dependencies: none

### How to test
For the Sphinx documentation,  go to `doc/` directory and follow [README](https://github.com/marian-nmt/marian-dev/tree/master/doc). For the Doxygen documentation, compile Marian with CMake option `-USE_DOXYGEN=ON`.

### Checklist

- [x] I have tested the code manually
- [ ] I have run regression tests
- [x] I have read and followed CONTRIBUTING.md
- [ ] I have updated CHANGELOG.md

### CODE REVIEW

File: doc/doc_guide.rst
Comment (snukky) [inline doc/doc_guide.rst:26]: Is this supposed to generate a box or something similar? In my local setup Sphinx generates a normal text from it. Is it correct? Maybe it should be a subsection?

File: doc/doc_guide.rst
Comment (snukky) [inline doc/doc_guide.rst:128]: Sphinx doesn't generate an image with math formula here for me and just keeps `|formula1|`. Is it supposed to do so?

File: doc/doc_guide.rst
Comment (snukky) [inline doc/doc_guide.rst:181]: These two missing parts are the main reason why this PR is a draft?

**Review by snukky — COMMENTED** [review]
Thank you @qianqianzhu for these changes! I rebased the branch with the recent master, updated instructions when I was setting this up, and simplified the documentation text a little. Let us know what you're planning to add to this PR before it's ready to be merged.

File: doc/requirements.txt
Comment (snukky) [inline doc/requirements.txt:5]: @qianqianzhu I had to decrease the version of myst-parser to 0.14 because later version have the requirement of Sphinx 3.0 or greater. Can you confirm what versions of Sphinx an myst-parser you use? Maybe we can increase the version for Sphinx to 3.x instead?

Comment (qianqianzhu) [reply doc/requirements.txt:5]: I can confirm that myst-parser 0.14 works fine for me. Thanks for the update.

File: doc/doc_guide.rst
Comment (qianqianzhu) [reply doc/doc_guide.rst:181]: Exactly. I am still working on it. I am quite new to restructureText, so it took longer than I expected. Hopefully I will complete those parts by this week.

Comment (snukky) [reply doc/doc_guide.rst:181]: Thanks. I don't think we need to write much in those sections. Probably a few sentences and a couple of links to external resources will be more than enough.

File: doc/doc_guide.rst
Comment (qianqianzhu) [reply doc/doc_guide.rst]: The image is missing. Now it should show as a inline image. Thanks for pointing out.

File: doc/doc_guide.rst
Comment (qianqianzhu) [reply doc/doc_guide.rst]: I guess it is now ready to review. I highlighted several syntaxes that are newly added to myst-parser. The reStructuredText is a bit complex to explain in a few sentence. So I leave it to Exhale and Sphinx's docs. Feel free to comment or commit changes.

**Review by snukky — APPROVED** [review]
Thank you @qianqianzhu. I slightly reorganized the structure and headers. If you are okay with the current state, I think we are ready to merge it.

**Comment by qianqianzhu** [comment]
Many thanks @snukky! Looks more smooth and sharp.

**Pull request merged**
